feat: introduce a character limit with a char limit counter UI

hydr063n opened the pull request:
Fixes https://github.com/status-im/nim-status-client/issues/1006

The chat input bar height adjusts to the message length (The height attributes supports integer so the transition isn't as smooth). The color of the message size changes from green to red if user has passed the character limit. The message won't be send if limit is passed, but instead a MessageModal will open explaining that to be courteous you need to adhere to 2000 char limit.

![](https://i.imgur.com/uvASeCQ.png)
![](https://i.imgur.com/apAI98Y.png)
![](https://i.imgur.com/YOnJyHG.png)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 0e809fc9 | [#1](https://ci.status.im/job/nim-status-client/job/macos/job/PR-1025/1/) | 2020-09-28 13:44:55 | ~7 min | `macos` | [:package:`dmg`](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-200928-133656-0e809f-pr.dmg) |

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        title: qsTrId("your-message-is-too-long.")
+        icon: StandardIcon.Critical
+        //% "Please make your message shorter. We have set the limit to 2000 characters to be courteous of others."
+        text: qsTrId("please-make-your-message-shorter.-we-have-set-the-limit-to-2000-characters-to-be-courteous-of-others.")
```
Oh thanks, that will indeed makes things so much easier! Ill be using that from now on!

hydr063n commented:
@jrainville  Ok I took a look at the files in `imports/theme` and will get acquainted with them so I can know where to use them. 

I also forgot to show an error when the user clicks the send button arrow as opposed to the the `return` key. So I've added this change in the commit I'm about to send now.

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        text: qsTrId("please-make-your-message-shorter.-we-have-set-the-limit-to-2000-characters-to-be-courteous-of-others.")
+        standardButtons: StandardButton.Ok
+        onVisibilityChanged: {
+            loading = false
```
We actually don't need that section of the code, I blindly copied the format of a message dialog from another qml. I need to pay more attention to these little details

hydr063n replied:
I've removed this section

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
     ChatButtons {
         id: sendBtns
-        height: parent.height
+        x: 451
```
No we don't, I will remove this bit

hydr063n commented:
@jrainville Ive just made all the changes you have requested.

jrainville commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
     ChatButtons {
         id: sendBtns
-        height: parent.height
+        width: 174
```
I don't think we need that width. `ChatButtons` should calculate its own width correctly

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        leftPadding: 8
+        anchors.verticalCenterOffset: -18
+
+        Text {
```
Ive made this change

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        anchors.right: parent.right
+        anchors.rightMargin: 15
+        anchors.verticalCenter: parent.verticalCenter
+        leftPadding: 8
```
Ive made this change

jrainville reviewed (commented):
Looks very good. I just think you could remove the `width` from ChatButtons since it should be handled by the component itself.
You can also test the difference between a public chat and a private chat. The private chat will have more buttons since you can send images directly and send chat commands. Those should still look ok

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        id: messageLengthLimitContainer
+        width: 174
+        anchors.right: parent.right
+        anchors.rightMargin: 15
```
Ive made this change

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        Text {
+            id: messageLengthLimit
+            text: (messageLimit - txtData.length).toString()
+            color: (messageLimit - txtData.length >= 0) ? "#32cd32" : "#ff0000"
```
Ive made this change

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
+        height: 36
         anchors.verticalCenter: parent.verticalCenter
         anchors.right: parent.right
+        anchors.verticalCenterOffset: 8
```
Ive made this change

hydr063n commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,52 @@ Rectangle {
         anchors.verticalCenter: parent.verticalCenter
         anchors.right: parent.right
+        anchors.verticalCenterOffset: 8
+        anchors.rightMargin: 15
```
Ive made this change

jrainville reviewed (approved):
Nice job

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,50 @@ Rectangle {
         }
     }
 
+    Row {
+        id: messageLengthLimitContainer
+        anchors.right: parent.right
+        anchors.rightMargin: Style.current.padding
+        anchors.verticalCenter: parent.verticalCenter
+        leftPadding: Style.current.halfPadding
+        anchors.verticalCenterOffset: -18
+
+        StyledText {
+            id: messageLengthLimit
+            text: (messageLimit - txtData.length).toString()
+            color: (messageLimit - txtData.length >= 0) ? Style.current.success : Style.current.danger
+            rightPadding: Style.current.halfPadding
+        }
+    }
```
This likely doesn't need a `Row`. A row is used to group multiple items together side-by-side without the need to anchor them to each other. Instead, you can move the anchors to the `StyledText` component instead.

emizzle replied:
It would be nice if we could add a `visible` property to this, that only appears once the message length is within 200 characters of the limit, similar to how Discord does it, ie:
```qml
property int messageLimitVisible: 200

StyledText {
    id: messageLimitLength
    visible: (messageLimit - txtData.length) <= root.messageLimitVisible 
}
```

emizzle replied:
What do you think about also anchoring this to the bottom right of the chat input, similar to how Discord does it? Additionally, we could move the chat buttons to the top. Thoughts @simonricoo?

![Imgur](https://imgur.com/6KgJo1R.png)
![Imgur](https://imgur.com/Sl5KeOg.png)

simonricoo replied:
@emizzle In figma its designed with the  new chat input. However you can see that the position is **bottom right**. Check figma for exact dimensions and padding :) 

Also I'm not sure if 

https://www.figma.com/file/Mr3rqxxgKJ2zMQ06UAKiWL/%F0%9F%92%AC-Chat%E2%8E%9CDesktop?node-id=2539%3A430

<img width="1413" alt="Screenshot_2020-09-29_at_09_24_59" src="https://user-images.githubusercontent.com/52249767/94526720-383ddd00-0236-11eb-9936-301c3216a046.png">

hydr063n replied:
> It would be nice if we could add a `visible` property to this, that only appears once the message length is within 200 characters of the limit, similar to how Discord does it, ie:
> 
> ```qml
> property int messageLimitVisible: 200
> 
> StyledText {
>     id: messageLimitLength
>     visible: (messageLimit - txtData.length) <= root.messageLimitVisible 
> }
> ```

This looks better than just going below 0. I've made the change

I've also removed the Row and only left the StyledText Component with the Row anchors/padding.

As per the figma I've anchored the buttons and char limit to the bottom right

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +289,50 @@ Rectangle {
+        StyledText {
+            id: messageLengthLimit
+            text: (messageLimit - txtData.length).toString()
+            color: (messageLimit - txtData.length >= 0) ? Style.current.success : Style.current.danger
```
What do you think about leaving the color as a secondary text color, and only making it red once the limit has been reached? Ie,
```qml
color: (messageLimit - txtData.length >= 0) ? Style.current.secondaryText : Style.current.danger
```
@simonricoo thoughts?

hydr063n replied:
Sounds good. I've made the change and also added a visible attribute that displays the character limit only when it has been surpassed.

emizzle reviewed (changes requested):
Very nice job @hydr063n! I've left this is as changes requested because I think we could improve slightly the positioning, visibility, and colouring of the character counter, with input from @simonricoo.

simonricoo commented:
@hydr063n Great job, one thing. Can we skip the positive indicator and show only the negative (red) number when you cross the limit, please?
<img width="802" alt="Screenshot_2020-09-29_at_09_32_22" src="https://user-images.githubusercontent.com/52249767/94527172-e21d6980-0236-11eb-984c-4c0da8639b79.png">

hydr063n commented:
Hey all, thanks so much for the help. I've anchored the buttons and char limit to the bottom right as per figma. I removed the Row component and only left the StyledText. I've also made the char limit to become visible only if the `messageLimitVisible` threshold has been surpassed.

![](https://i.imgur.com/YfUtqi2.png)
![](https://i.imgur.com/AGoOIo6.png)
![](https://i.imgur.com/VKO8A3i.png)

jrainville commented:
Just an FYI, @PascalPrecht is refactoring the ChatInput a lot at the same time, so making big  refactors here is useless since the Input itself will change a lot

PascalPrecht commented:
I think it's fine. If my refactor takes longer to land, then that's what it is. I'll rebase my work on top of this then

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -276,18 +290,45 @@ Rectangle {
     }
 
+    StyledText {
+        id: messageLengthLimit
+        text: (messageLimit - txtData.length).toString()
+        visible: (messageLimit - txtData.length) <= root.messageLimitVisible
+        color: ((messageLimit - txtData.length) <= 0) ? Style.current.danger : Style.current.textColor
```
```suggestion
        id: messageLengthLimit
        property int remainingChars: messageLimit - txtData.length
        text: remainingChars.toString()
        visible: remainingChars <= root.messageLimitVisible
        color: (remainingChars <= 0) ? Style.current.danger : Style.current.textColor
```

emizzle reviewed (commented):
Needs a rebase, and one last remaining suggestion 😄

emizzle commented:
Also, don't forget to squash your commits 😄

hydr063n commented:
hey @emizzle  :smile: I've added the new property variable. That really makes the section more readable! I also squashed and rebased the work on top of today's master (as of writing) :rocket:

jrainville commented on ui/app/AppLayouts/Chat/ChatColumn/ChatInput.qml:
```diff
@@ -63,12 +72,17 @@ Rectangle {
-                event.accepted = true;
-                return
+            if (txtData.length < messageLimit) {
+                if (emojiSuggestions.visible) {
+                    emojiSuggestions.addEmoji();
+                    event.accepted = true;
+                    return
+                }
```
This `if` can be put outside the `if (txtData.length < messageLimit) {` since putting a new emoji should be accepted whether or not the limit is reached

hydr063n replied:
Whoops, just fixed this!

hydr063n replied:
Whoops, just fixed this!
